TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/util/ActiveContextMonitor.java

```diff
@@ -0,0 +1,263 @@
+        this.orderedRequestTracker = orderedRequestTracker;
+        this.totalItemsToOutputLimit = totalItemsToOutputLimit;
+        this.logger = logger;
+        this.formatWorkItem = formatWorkItem;
+        this.logLevelIsEnabled = logLevelIsEnabled;
+        ageToLevelEdgeMapRef = new AtomicReference<>();
+        setAgeToLevelMap(levelShowsAgeOlderThanMap);
+    }
+
+    public void setAgeToLevelMap(Map<Level, Duration> levelShowsAgeOlderThanMap) {
+        ageToLevelEdgeMapRef.set(levelShowsAgeOlderThanMap.entrySet().stream()
+                .collect(Collectors.toMap(Map.Entry::getValue, Map.Entry::getKey,
+                        (x,y) -> { throw Lombok.sneakyThrow(new IllegalStateException("Shouldn't have any merges")); },
+                        TreeMap::new)));
+    }
+
+    Duration getAge(long recordedNanoTime) {
+        return Duration.ofNanos(System.nanoTime()-recordedNanoTime);
+    }
+
+    /**
+     * Try to print out the most valuable details at the end, assuming that a user is tailing a file that's
+     * constantly being appended, and therefore could be harder to home in on the start of a block.
+     */
+    public void logTopOpenActivities() {
+        logRequests().ifPresent(ll->logger.accept(ll, ()->"\n"));
+        logTopActiveScopes().ifPresent(ll->logger.accept(ll, ()->"\n"));
+        logTopActiveScopesByType().ifPresent(ll->logger.accept(ll, ()->"\n"));
+    }
+
+    public Optional<Level> logTopActiveScopesByType() {
+        return perActivityContextTracker.getActiveScopeTypes().map(c->
+                        gatherActivities(perActivityContextTracker.getOldestActiveScopes(c),
+                                perActivityContextTracker.numScopesFor(c),
+                                this::getLogLevelForActiveContext))
+                .sorted(Comparator.comparingDouble(ActivitiesAndDepthsForLogging::getAverageContextDepth))
+                .map(cad-> {
+                    if (cad.items.isEmpty()) { return Optional.<Level>empty(); }
+                    final var sample = cad.items.get(0);
+                    logger.accept(getHigherLevel(Optional.of(sample.getValue()), Optional.of(Level.INFO)).get(), () ->
+                            OLDEST_ITEMS_FROM_GROUP_LABEL_PREAMBLE + cad.totalScopes + " scopes that are past " +
+                                    "thresholds for '" + sample.getKey().getActivityName() + "'");
+                    cad.items.forEach(kvp->logger.accept(kvp.getValue(), ()->activityToString(kvp.getKey())));
+                    return (Optional<Level>) Optional.of(cad.items.get(0).getValue());
+                })
+                .collect(Utils.foldLeft(Optional.<Level>empty(), ActiveContextMonitor::getHigherLevel));
+    }
+
+    private static Optional<Level> getHigherLevel(Optional<Level> aOuter, Optional<Level> bOuter) {
+        return aOuter.map(a -> bOuter.filter(b -> a.toInt() <= b.toInt()).orElse(a))
+                .or(() -> bOuter);
```

Comment:
All we're doing here is comparing the int values of each level, right? It seems like a bewilderingly complex line to just do that

Reply:
Sortof - the types are Optional<Level> where Level is an enum (hence the toInt()).  The Optional parts require the map and two ors.  If I had a monad on Optional to say, take the one that was present OR do this comparison, that would be cleaner, but I don't know of one